### PR TITLE
Page tracking

### PR DESCRIPTION
**Pull request by Miniplop**
## Types of changes

- [ ] Bug fix (a non-breaking change which fixes an issue)
- [ ] New feature (a non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [ ] All new and existing tests are passing.

### CODE REVIEW

File: packages/slice-machine/pages/_app.tsx [hunk 1]
```diff
@@ -74,77 +42,38 @@ const RemoveDarkMode: React.FunctionComponent = ({ children }) => {
...
-      Object.keys(newSliceMap).forEach((key) => {
-        // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
-        if (!sliceMap[key]) {
-          return (window.location.href = `/slices`);
-        }
-      });
-    }
-    setSliceMap(newSliceMap);
-    const { env, configErrors, warnings, libraries } = serverState;
-    console.log("------ SliceMachine log ------");
```
Comment (MarcMcIntosh) [inline packages/slice-machine/pages/_app.tsx:143]: we're removing the logs?

File: packages/slice-machine/server/src/api/index.ts [hunk 2]
```diff
@@ -64,11 +64,7 @@ router.get(
   ): Promise<Express.Response> {
     // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-call
     const payload = await state(req);
-    // eslint-disable-next-line @typescript-eslint/no-unsafe-member-access
-    if (payload.clientError) {
```
Comment (MarcMcIntosh) [inline packages/slice-machine/server/src/api/index.ts:68]: how are errors sent to the client now?

File: packages/slice-machine/src/apiClient.ts [hunk 3]
```diff
@@ -17,6 +18,12 @@ const defaultAxiosConfig = {
   },
 };
 
+/** State Routes **/
+
+export const getState = (): Promise<AxiosResponse<ServerState>> => {
+  return axios.get("/api/state", defaultAxiosConfig);
```
Comment (MarcMcIntosh) [inline packages/slice-machine/src/apiClient.ts]: you can pass `ServerState` as a type parameter to `axios.get` not sure if it makes a difference here, as the response object is kept

File: packages/slice-machine/src/hooks/useSMTracker.tsx [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  getCurrentVersion,
+  getFramework,
+  getIsTrackingAvailable,
+  getRepoName,
+  getShortId,
+} from "@src/modules/environment";
+import { getLibraries } from "@src/modules/slices";
+import { useRouter } from "next/router";
+
+const useSMTracker = () => {
```
Comment (MarcMcIntosh) [inline packages/slice-machine/src/hooks/useSMTracker.tsx:16]: neat

File: packages/slice-machine/src/hooks/useServerState.tsx [hunk 5]
```diff
@@ -2,20 +2,23 @@ import { useEffect } from "react";
 
 import ServerState from "@lib/models/server/ServerState";
 import useSliceMachineActions from "@src/modules/useSliceMachineActions";
+import useSwr from "swr";
+import axios, { AxiosResponse } from "axios";
 
-const useServerState = (serverState: ServerState) => {
-  const { getState } = useSliceMachineActions();
+const fetcher = (url: string): Promise<ServerState> =>
+  axios.get(url).then((res: AxiosResponse<ServerState>) => res.data);
```
Comment (MarcMcIntosh) [inline packages/slice-machine/src/hooks/useServerState.tsx:9]: `axios.get<ServerState>` and the res will be inferred for you :)

File: packages/slice-machine/src/hooks/useServerState.tsx [hunk 6]
```diff
@@ -2,20 +2,23 @@ import { useEffect } from "react";
...
-    serverState.configErrors,
-    serverState.customTypes,
-    serverState.remoteCustomTypes,
-    serverState.libraries,
-  ]);
+    if (!serverState) {
+      return;
+    }
+
+    refreshState(serverState);
```
Comment (MarcMcIntosh) [inline packages/slice-machine/src/hooks/useServerState.tsx:20]: what's refresh state do?

File: packages/slice-machine/src/tracker.ts [hunk 7]
```diff
@@ -63,34 +57,41 @@ export class SMTracker {
...
 
-  async #group(attributes: Record<string, unknown> = {}): Promise<void> {
+  async #identify(userId: string): Promise<void> {
     if (!this.#isTrackingPossible(this.#client)) {
       return;
     }
 
-    const repoName = this.#repoName;
+    return this.#client
+      .then((client): void => {
```
Comment (MarcMcIntosh) [inline packages/slice-machine/src/tracker.ts:77]: why void?
plus if you add void before the `client.identify` promise the warning should go away

File: packages/slice-machine/src/tracker.ts [hunk 8]
```diff
@@ -63,34 +57,41 @@ export class SMTracker {
...
+  async #group(attributes: Record<string, unknown> = {}): Promise<void> {
+    if (!this.#isTrackingPossible(this.#client)) {
       return;
     }
 
     return this.#client
       .then((client): void => {
         // eslint-disable-next-line @typescript-eslint/no-floating-promises
-        client.group(repoName, attributes);
+        client.group(attributes);
```
Comment (MarcMcIntosh) [inline packages/slice-machine/src/tracker.ts]: you can add void here to :)

**Review by MarcMcIntosh — COMMENTED** [review]
I have a few questions and recommendations :)

File: packages/slice-machine/server/src/api/index.ts (hunk 2 above)
Comment (Miniplop) [reply packages/slice-machine/server/src/api/index.ts:68]: The error are not sent anymore but in the client side we never use the fact that this call can send a different status code that's why I removed it.

File: packages/slice-machine/src/hooks/useServerState.tsx (hunk 6 above)
Comment (Miniplop) [reply packages/slice-machine/src/hooks/useServerState.tsx:20]: send the data coming from the /api/state into the redux store

File: packages/slice-machine/pages/_app.tsx (hunk 1 above)
Comment (Miniplop) [reply packages/slice-machine/pages/_app.tsx:143]: did you use it ? I never used it

File: packages/slice-machine/src/apiClient.ts (hunk 3 above)
Comment (Miniplop) [reply packages/slice-machine/src/apiClient.ts]: The type we put in the `axios.get` function is only use on `then`, `catch`etc ... functions, but doesn't help typescript to infer the return type of the function, so now I put both :)

File: packages/slice-machine/src/hooks/useSMTracker.tsx (hunk 4 above)
Comment (Miniplop) [reply packages/slice-machine/src/hooks/useSMTracker.tsx:16]: 😎

File: packages/slice-machine/src/hooks/useServerState.tsx (hunk 5 above)
Comment (Miniplop) [reply packages/slice-machine/src/hooks/useServerState.tsx:9]: Typescript can't inferred the `then` type if I use the generics, don't know why

File: packages/slice-machine/src/tracker.ts (hunk 7 above)
Comment (Miniplop) [reply packages/slice-machine/src/tracker.ts:77]: Because this function return nothing, just calling the segment segment.

Comment (Miniplop) [reply packages/slice-machine/src/tracker.ts:77]: I add the void on the promise to remove the eslint comment

**Review by MarcMcIntosh — APPROVED** [review]

**Pull request merged**